### PR TITLE
Fix dataloaders are not reset when tuning the model

### LINKED ISSUE #7562: stopped calling datamodule.train_dataloader() when tuning the batch_size using pl.trainer.tune()
**version 1.3.1 pytorch_lightining stopped calling datamodule.train_dataloader() when tuning the batch_size using pl.trainer.tune()**

I don't know if this was reported already, but since version 1.3.1 pytorch_lightining stopped calling datamodule.train_dataloader() when tuning the batch_size using pl.trainer.tune(), so the batch_size effectivelly stays fixed at 2, see the attached files, in particular the print statements of

```python
    def train_dataloader(self):
        print("THIS SHOULD BE CALLED!!!!!", self.batch_size)
```
and
```python
    def training_step(self, train_batch, batch_idx):
        inputv, target = train_batch
        output = self.forward(inputv)
        print("actual batch_size", len(inputv))
```
when tune is called.

[code_examples.zip](https://github.com/PyTorchLightning/pytorch-lightning/files/6488147/code_examples.zip)

### PR DESCRIPTION
**Pull request by Lucklyric**
## What does this PR do?

Fixes #7562 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -188,10 +189,10 @@ def reset_train_val_dataloaders(self, model) -> None:
         The val dataloader must be initialized before training loop starts, as the training loop
         inspects the val dataloader to determine whether to run the evaluation loop.
         """
-        if self.trainer.train_dataloader is None:
+        if self.trainer.train_dataloader is None or self.trainer.state.status == TrainerStatus.RUNNING:
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:192]: The `TrainerStatus` should always be `RUNNING` here. What you added is equivalent to removing the `if` so definitely not the solution

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -188,10 +189,10 @@ def reset_train_val_dataloaders(self, model) -> None:
...
             self.trainer.reset_train_dataloader(model)
 
-        if self.trainer.val_dataloaders is None:
+        if self.trainer.val_dataloaders is None or self.trainer.tuning:
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:195]: if this fix works, it tells me that the problem is somewhere else

shouldn't have to check this here.

**Review by carmocca — CHANGES_REQUESTED** [review]
Can you add a test that fails in current master and is fixed by the changes in this PR?

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (Lucklyric) [reply pytorch_lightning/trainer/training_loop.py:195]: Sorry, I thought it is "TunerStatus". Yes, you are right, it should always RUNNING. In fact that, I had one solution that adding `or self.trainer.tuning` for both train and val dataloaders. And switch 

self.trainer.training = True       
self.trainer._run(*args, **kwargs) 
self.trainer.tuning = True

to

self.trainer.training = True       
self.trainer.tuning = True
self.trainer._run(*args, **kwargs) 

in

https://github.com/PyTorchLightning/pytorch-lightning/blob/20f63377f81f4771d3f128f979b3a0f9b8d219a7/pytorch_lightning/tuner/tuning.py#L60-L65
But it breaks one existing test in test_model_state
https://github.com/PyTorchLightning/pytorch-lightning/blob/20f63377f81f4771d3f128f979b3a0f9b8d219a7/tests/trainer/test_states.py#L50
Do you think we should keep this direction to fix this problem and work on solving the test issue? Because we need to reset the dataloder for each `tuner._run()` -> `trainer._run()`. Because the dataloader is attached only once before the tuning, hence we need to reset the dataloder even it is not none when it is in tuning status.

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:195]: > Do you think we should keep this direction to fix this problem and work on solving the test issue? 

No, that fix would also not be correct. Due to how tuning works, we need to set the trainer stage to training before the tuner calls `_run` internally.

If you can add the failing test that covers the bug, I can have a look and try to resolve it.

Comment (Lucklyric) [reply pytorch_lightning/trainer/training_loop.py:195]: Yes, I understand that. Because the model stage can only be either Runningtage.TUNING or RunningStage.TRAINING. 
Hence, if I move `tuning = True` before `_run`, it results in the property of `trainer.training` becomes False.  
I think we need somehow let logic inside `trainer._run` know it is called from tunner and reset the dataloader's. Or we need to find a way to reset the dataloader before calling `_run` in Tuner. What do you think?

Comment (Lucklyric) [reply pytorch_lightning/trainer/training_loop.py:195]: I am thinking, what about just simply call reset_train_dataloder in `tuner._run`? 
```python
   def _run(self, *args: Any, **kwargs: Any) -> None:
        """`_run` wrapper to set the proper state during tuning, as this can be called multiple times"""
        self.trainer.state.status = TrainerStatus.RUNNING  # last `_run` call might have set it to `FINISHED`
        self.trainer.training = True
        self.trainer.reset_train_dataloader(model)
        self.trainer._run(*args, **kwargs)
        self.trainer.tuning = True
```

Comment (Lucklyric) [reply pytorch_lightning/trainer/training_loop.py:195]: Can we check if the `trainer.train_loop.global_step == 0` in `reset_train_val_dataloaders()`

Comment (Lucklyric) [reply pytorch_lightning/trainer/training_loop.py:195]: Hi @carmocca , what do you think if we simply reset `trainer.train_dataloader` to None at each interaction step when searching for batch size? 
```python
def _run_power_scaling(
    trainer: 'pl.Trainer', model: 'pl.LightningModule', new_size: int, batch_arg_name: str, max_trials: int
) -> int:
    """ Batch scaling mode where the size is doubled at each iteration until an OOM error is encountered. """
    for _ in range(max_trials):
        garbage_collection_cuda()
        trainer.train_loop.global_step = 0  # reset after each try
        try:
            #!!!! Set trainer train_dataloader to None in order to reset it
            trainer.train_dataloader = None
            # Try fit
            trainer.tuner._run(model)
            # Double in size
            new_size, changed = _adjust_batch_size(trainer, batch_arg_name, factor=2.0, desc='succeeded')
        except RuntimeError as exception:
            # Only these errors should trigger an adjustment
            if is_oom_error(exception):
                # If we fail in power mode, half the size and return
                garbage_collection_cuda()
                new_size, _ = _adjust_batch_size(trainer, batch_arg_name, factor=0.5, desc='failed')
                break
            else:
                raise  # some other error not memory related

        if not changed:
            break
    return new_size
```
Same for `_run_binsearch_scaling`.

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:195]: Yes, that seems like a better option to me.

```python
if changed:
    # set dataloaders to None so they are reset
else:
    break

Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py]: @carmocca shouldn't it be:

```
for _ in range(max_trials):
    garbage_collection_cuda()
    trainer.train_loop.global_step = 0  # reset after each try
    try:
        # Try fit
        trainer.tuner._run(model)
        # Double in size
        new_size, changed = _adjust_batch_size(trainer, batch_arg_name, factor=2.0, desc='succeeded')
    except RuntimeError as exception:
        # Only these errors should trigger an adjustment
        if is_oom_error(exception):
            # If we fail in power mode, half the size and return
            garbage_collection_cuda()
            new_size, _ = _adjust_batch_size(trainer, batch_arg_name, factor=0.5, desc='failed')
            break
        else:
            raise  # some other error not memory related

    if not changed:
        break

if changed:
    trainer.train_dataloader = None

return new_size
```

**Review by tchaton — APPROVED** [review]
LGTM ! Good catch !

**Review by carmocca — APPROVED** [review]
Added the test that fails in current master myself. LGTM

**Review by carmocca — CHANGES_REQUESTED** [review]
Just noticed I'm not able to push changes to your fork's branch. I'm guessing you have disabled this:

https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Can you enable it? Alternatively, you can cherry-pick this commit https://github.com/PyTorchLightning/pytorch-lightning/commit/d88b88c465203f1919737eeb35c94493db6cfb9e

**Comment by Lucklyric** [comment]
Dear @carmocca , it seems the option is already enabled.  NVM, I have cherry-picked your revision.  Thank you so much.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/tuner/batch_size_scaling.py [hunk 3]
```diff
@@ -192,7 +195,10 @@ def _run_binsearch_scaling(
             else:
                 new_size, changed = _adjust_batch_size(trainer, batch_arg_name, factor=2.0, desc='succeeded')
 
-            if not changed:
+            if changed:
+                # Force the train dataloader to reset as the batch size has changed
+                trainer.train_dataloader = None
```
Comment (awaelchli) [inline pytorch_lightning/tuner/batch_size_scaling.py:200]: what's the reason for not calling the reset method?

Comment (carmocca) [reply pytorch_lightning/tuner/batch_size_scaling.py:200]: I guess that would work too. Let's change it.

Comment (Lucklyric) [reply pytorch_lightning/tuner/batch_size_scaling.py]: Dear @awaelchli , thanks for your suggestion. Updated!

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Review by awaelchli — APPROVED** [review]
